pkg/resources/grant_ownership.go

```diff
@@ -0,0 +1,668 @@
+				Detail:   fmt.Sprintf("Id: %s\nError: %s", d.Id(), err),
+			},
+		}
+	}
+
+	grantOn, err := getOwnershipGrantOn(d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if grantOn.Future != nil {
+		// TODO (SNOW-1182623): Still waiting for the response on the behavior/SQL syntax we should use here
+	} else {
+		accountRoleName, err := client.ContextFunctions.CurrentRole(ctx)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		err = client.Grants.GrantOwnership( // TODO: Should we always set outbound privileges to COPY in delete operation or set it to the config value?
+			ctx,
+			grantOn,
+			sdk.OwnershipGrantTo{
+				AccountRoleName: sdk.Pointer(sdk.NewAccountObjectIdentifier(accountRoleName)), // TODO: What if current role is database role (not a common but possible case)
+				// DatabaseRoleName: TODO: handle later
+			},
+			getOwnershipGrantOpts(id),
+		)
+		if err != nil {
+			return diag.Diagnostics{
+				diag.Diagnostic{
+					Severity: diag.Error,
+					Summary:  "An error occurred when transferring ownership back to the original role",
+					Detail:   fmt.Sprintf("Id: %s\nError: %s", d.Id(), err),
+				},
+			}
+		}
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func ReadGrantOwnership(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+	id, err := ParseGrantOwnershipId(d.Id())
+	if err != nil {
+		return diag.Diagnostics{
+			diag.Diagnostic{
+				Severity: diag.Error,
+				Summary:  "Failed to parse internal identifier",
+				Detail:   fmt.Sprintf("Id: %s\nError: %s", d.Id(), err),
+			},
+		}
+	}
+
+	opts, grantedOn := prepareShowGrantsRequestForGrantOwnership(id)
+	if opts == nil {
+		return nil
+	}
+
+	client := meta.(*provider.Context).Client
+
+	grants, err := client.Grants.Show(ctx, opts)
+	if err != nil {
+		return diag.Diagnostics{
+			diag.Diagnostic{
+				Severity: diag.Error,
+				Summary:  "Failed to retrieve grants",
+				Detail:   fmt.Sprintf("Id: %s\nError: %s", d.Id(), err),
+			},
+		}
+	}
+
+	ownershipFound := false
+
+	for _, grant := range grants {
+		if grant.Privilege != "OWNERSHIP" {
+			continue
+		}
+
+		// Future grants do not have grantedBy, only current grants do.
+		// If grantedby is an empty string, it means terraform could not have created the grant
+		if (opts.Future == nil || !*opts.Future) && grant.GrantedBy.Name() == "" {
+			continue
+		}
+
+		// grant_on is for future grants, granted_on is for current grants.
+		// They function the same way though in a test for matching the object type
+		if grantedOn != grant.GrantedOn && grantedOn != grant.GrantOn {
+			continue
+		}
+
+		switch id.GrantOwnershipTargetRoleKind {
+		case ToAccountGrantOwnershipTargetRoleKind:
+			if grant.GranteeName.Name() == id.AccountRoleName.Name() {
+				ownershipFound = true
+			}
+		case ToDatabaseGrantOwnershipTargetRoleKind:
+			if grant.GranteeName.Name() == id.DatabaseRoleName.Name() {
+				ownershipFound = true
+			}
+		}
+	}
+
+	if !ownershipFound {
```

Comment:
so should it be an error? in other (normal) resources, if we don't find the object we mark it as removed based on the error. So why is the no ownership an error. In the following example:
1. config with ownership
2. manual ownership transfer
3. terraform plan

the result would be an error. Should it be an error or should we try to set the ownership with tf?

Reply:
As discussed, we should set id to an empty string, and in the create operation we should try to forcefully take ownership and then grant it (or just grant it straight to the target role). Maybe to make it simpler I'll add it to the list of things I'll do in the next pr (should be added to the pr desc).